### PR TITLE
refactor per #503, fixing supervise and supervise_as handling of args

### LINKED ISSUE #503: Registry not working for SupervisionGroups via supervise
I have a situation where I am customizing the behavior and details of a SupervisionGroup. Using pools works just fine, but if I want to supervise only one actor, it appears as though the registry isn't working correctly. 
```
@supervisor = Celluloid::SupervisionGroup.run!

# works
@supervisor.pool(SocialData::Managers, as: :data_managers, size: 2, args: [data_name: "twitter"])
@supervisor[:data_managers].process 

# returns nil:NilClass (NoMethodError)
@supervisor.supervise(SocialData::Manager, as: :data_manager, args: [data_name: "twitter"])
@supervisor[:data_manager].process 
```
Based on everything I see, I should be able to use the above syntax to supervise one actor via a supervision group. The Celluloid::Actors approach also doesn't show the actor in the registry. Is this intentional? I am currently working with the latest from Master.

### PR DESCRIPTION
**Pull request by digitalextremist**
Fixes #503

### CODE REVIEW

File: lib/celluloid/supervision_group.rb [hunk 1]
```diff
@@ -123,6 +123,7 @@ def initialize(registry, klass, options = {})
         options = options.each_with_object({}) { |(k,v), h| h[k.to_s] = v }
 
         @name = options['as']
+        puts "name? #{@name}"
```
Comment (tarcieri) [inline lib/celluloid/supervision_group.rb]: Looks good other than this...

**Comment by digitalextremist** [comment]
@tarcieri good catch, artifact removed. Merging and closing #503 also.

**Comment by digitalextremist** [comment]
( build failed due to branch delete )

**Comment by e2** [comment]
@digitalextremist - please don't merge failing code into celluloid/master - I know waiting for Travis can be a PITA (especially if the changes seem harmless like this one), but at least run the specs locally before you create/update the PR.

There's simply too much "magic" happening in Celluloid (`:respond_to?`, `:send`) - so you never know what you could really be breaking, while investigating any failure can be a massive time sink.

Also, I need to use unreleased patches merged to master, and I don't want to lock to a specific commit - but a bad merge like this breaks PRs for libraries using celluloid, e.g. https://travis-ci.org/guard/listen/jobs/55545343

This commit is especially bad, because it causes the Celluloid::IncidentReporter to crash (rendering Celluloid completely useless, no matter how it's used).

**Comment by digitalextremist** [comment]
@e2, I am sorry you had trouble, and I agree local testing is crucial, which is what I do, and did here. Travis is watched, but also the same set of rubies can be locally raked. But like you said, there's a lot of voodoo going on so short of regression testing every dependent library there's going to be a snag sometimes. I'm not sure why you'd not expect that considering this is master we're talking about. 

Are you catching the commits before or after the `autostart` fix? This piece of code is part of a pretty dangerous and critical overhaul involving breaking changes. If you're not mapping to a `-stable` branch, here and in general you're going to be facing a rough transition until we fully cross this transition. Many libraries depend on Celluloid as you said, and I depend on those not to break because they're expecting `master` to be perfect. In addition to using a stable branch, in my own work I always work off a fork of my own and never the root repository. All this said, what can I do to help?

**Comment by digitalextremist** [comment]
And I'm further confused as this has been resolved and quickly at that in #505.

**Comment by e2** [comment]
@digitalextremist - don't worry about it, I'm just asking you to be more careful, because people rely on celluloid/master more than you probably think (how would you otherwise know, right?).

> there's going to be a snag sometimes

In this case I checked out celluloid at eb5f362 and the unit tests didn't work locally. I'd understand mistakes such as if you accidentally merged a wrong branch/etc. But allowing broken unit tests on master is one of the most detrimental things I can think of (in this context).

It's not about regression testing - it's about allowing Celluloid to work at all here (as I mentioned - the IncidentReporter crashes Celluloid on startup in this commit because of the STDERR argument passed to `Celluloid.supervise_as`).

> I'm not sure why you'd not expect that considering this is master we're talking about.

The master branch is for people to cooperate on - I'm sure other Celluloid contributors will want master working most of the time (it would be ok if you were the only person in the world hacking on Celluoid and preparing contributions - otherwise it's best to have master working at all times).

If you want to "test stuff" - that's what you create a separate branch for (so master is always ready for unexpected bugfixes) - and you can tell people to test that branch (instead of "try master now").

Sure, master doesn't mean "production ready", but then again it doesn't mean "completely broken and failing unit tests" - that's the reason Travis exists, right?.

> Are you catching the commits before or after the autostart fix?

It was broken for 3 days I think. The best action is usually to revert the fix so that master works again.

> This piece of code is part of a pretty dangerous and critical overhaul involving breaking changes. 

And there are no tests for the changes you made: there should be spec examples with single non-hash arguments. Just like the way the IncidentReporter is supervised in the actor system.

If it's that dangerous, there should be tests, right?

> If you're not mapping to a -stable branch

There's no 0-16-stable branch! There's only 0-15-stable, which is 2 years old. Typically, it's reasonable to expect the master branch to be `stable enough` - especially if that's where contributions are supposed to be merged, right?

> you're going to be facing a rough transition until we fully cross this transition

The do it on a separate branch until it's "ready for testing". Broken unit tests is not "ready for testing". Otherwise, it would be reasonable for me to expect you to create a 0-16-stable branch and backport #497 (because I need it for development on projects using RSpec > 3) - but I don't know why you'd want people to bother you with backport patches just so you can make breaking changes on master.

> In addition to using a stable branch, in my own work I always work off a fork of my own and never the root repository

I (almost) always work of the master branch of any project if I want to contribute any bugfixes - because that's the branch the PRs will be merged into, right?

What if there's an important bug to fix while you're "half way" doing the breaking changes? How would you prepare a release?

> All this said, what can I do to help?

1. keep all unit tests working on master at all times - not unreasonable, especially because of PRs - and because you can release a new version at any time with little risk
2. track coverage and add unit tests as needed - I know this is tricky in Celluloid, but it's the only way to make compatibility-breaking changes without creating core-feature-breaking changes
3. if you are making many changes over many commits over a long period of time, use a separate branch - because it will be easier for you to rebase to master than for other people to rebase their fixes on top of your breaking changes (the alternative is to ask for contributions onto stable branches, but this is a lot more work for everyone with no good reason)

**Comment by tarcieri** [comment]
@e2 people shouldn't be relying on master. They should be relying on `0-16-stable`. If there are unreleased patches you need on that branch, let me know. I can also spin another release.

**Comment by e2** [comment]
@tarcieri - 0-16-stable didn't exist last time I checked.

>  people shouldn't be relying on master

I know that - this is purely for development and testing for possible issues before releases. Though, Celluloid doesn't even have a single major release (that way I could just stick to e.g. ~1.x and forget about it completely until Celluloid 2.0.0 comes out) and people probably have project using 0.15.x and 0.16.x so I don't want to force either.

Development in Celluloid isn't that fast either, and I don't mind to get my hands dirty if celluloid/master breaks something due to improvements or bugfixes - as long as there isn't a huge difference between the latest release and master.

Ideally for development and tests I'd want to test against any "pre-release" gems in the future - does locking to `'~ 0.x'` make sense? Given there are no plans for 1.0.0 AFAIK, and since I didn't see a 0-16-stable previously, it made sense to me to use master.

Sorry - I'm just a bit confused here about what to expect in the future. And having an unstable master discourages me from contributing (because multiple people can end up wasting time investigating the same problems).

> If there are unreleased patches you need on that branch, let me know. 

Nothing worth the backporting really (I respect your time) - just @11c1060 AFAIK (otherwise it means changing RSpec configurations - just for the sake of Celluloid test helpers).

I don't expect master to be "production ready" or even work reliably - but if you consider basic unit tests failing on master to be "ok", let me know and I'll just lock to 0.16.x everywhere and stop testing anything newer until maybe 0.17.x or even 1.0.0.

**Comment by digitalextremist** [comment]
@e2 please use the Google group for the future if you have a complaint or are going to put so much non-resolution oriented text into it. We keep basically every topic you hit today in discussion boards. You've said a lot, your opinions you are welcome to have, but we keep a clear workspace.

**Comment by tarcieri** [comment]
@e2 I just pushed the current master to `0-16-stable`. Can you confirm it needs your needs? If so I can cut 0.16.1

**Comment by e2** [comment]
@tarcieri - Thanks! Works perfect!

@digitalextremist - sorry about that. I'm just worried, because often whenever unit tests are no longer required to pass on master, the given project quickly turns to shit. And I just don't want Celluloid turning to shit - thus my complaining. Breaking the API Celulloid uses internally itself, not treating it as an accident, and not covering it with unit tests, telling people to "test it on master" anyway, then saying master shouldn't be relied on ... it's like "abandon all hope all ye who depend on Celluloid". I thought this was a single incident, but I'll make sure to keep topics like this within Google Groups - just in case.

**Comment by digitalextremist** [comment]
@e2 if that's where you're coming from, you've got a friend in me. You came on strong & hand-slappy, and I had to bite my tongue on accusations made, but I definitely one of your allies. Real fast since we're here, other than that I'd continue by Google Groups:

I can personally assure you Celluloid is going no where but forward. It's got active, very active people, a long bright future, and people who are truly devoted. I was brought into the three-person core group with a strong conviction that Celluloid needs to hit `1.0` soon, broaden in audience/community, and do a lot that's in its destiny so-to-speak. I will *always* discuss anything that is bothering you, as @tarcieri has also proven time and again, but when you come at us like we killed your first born child it tends to dampen things. And here you came at us guerrilla because you feel the same way we do. Who knew?

Final thing: @tarcieri answered the tests issue. The API breaking is part of why we've not released `1.0` yet, to reserve the right to break Celluloid to near death and refactor it ( but we try not to ). I'm a believer in the "broken window" concept where a city-block will remain pristine until the first broken window, then next thing you know there are transients living in boarded houses and stuff is burning. What you've seen is no broken window, it's an unfinished universe... The Celluverse is still rendering.

**Comment by e2** [comment]
@digitalextremist - thanks, it's a spot-on summary and is much appreciated. I consider you guys pros, so don't take anything I say personally. There's a crap load of work that has been (and will be) put into Celluloid that never will be appreciated (because few people even can) - just continue being a pro (it's tough given the pressure, I know) and keep up the great work! Thanks!

**Comment by ioquatix** [comment]
Um, is it just me or is this ugly:

```ruby
@supervisor = TestServer.supervise(args: [{listen: SERVER_PORTS}])
@supervisor.actors.first.run
```

**Comment by digitalextremist** [comment]
@ioquatix: celluloid/celluloid-supervision#18

**Pull request merged**